File: 011. 盛最多水的容器.py
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        # 如果你移动长板：宽度变小，而高度受限于原有的短板，面积一定会减小。
        # 如果你移动短板：宽度虽然变小，但有可能遇到更长的板子，从而提升高度，面积有可能增大。
        n = len(height)
        left = 0
        right = n - 1
        max_volume = (right - left) * min(height[left] , height[right])
        while right > left:
            if height[left] > height[right]:
                right -= 1
            else:
                left += 1
            volume = (right - left) * min(height[left] , height[right])
            max_volume=max(max_volume,volume)
        return max_volume
s = Solution()
print(s.maxArea([1,8,6,2,5,4,8,3,7]))

File: 011. 盛最多水的容器.go
package main

func maxArea(height []int) int {
	// 双指针初始指向两端，谁的高度矮，谁就往里缩,每缩一次都要维护最大水量
	var res int
	l := 0
	r := len(height) - 1

	for l < r {
		if height[l] <= height[r] {
			water := (r - l) * height[l]
			if water > res {
				res = water
			}
			l++
		} else {
			water := (r - l) * height[r]
			if water > res {
				res = water
			}
			r--
		}
	}
	return res
}
